### DISCUSSION TITLE
[OneShotSyncAgent] NegotiatorInfo may disappear even when it is the beginning of negotiations

### DISCUSSION THREAD

**Discussion by a4rcvv** [yasserfarouk/scml#117] [one-shot-qa] [answered]
I'm implementing an agent which extends `OneShotSyncAgent`. Sometimes (but not always), `NegotiatorInfo` of some agent disappears from `self.negotiators` even when it is the very beginning of negotiations.

For example, I implement `first_proposals()` as follows. The agent uses `self.get_nmi(partner_id)` in `self.get_proposals()`.

```python
def first_proposals(self) -> dict[str, Outcome]:
    try:
        return self.get_proposals(set(self.partners()))
    except Exception as e:
        print(f"negotiator: {self.negotiators}")
        print(f"supplier: {self.awi.all_suppliers} ")
        print(f"consumer: {self.awi.all_consumers}")
        raise e
```

The output says:

```
negotiator: {'05SAS@1': NegotiatorInfo(negotiator=<negmas.sao.negotiators.controlled.ControlledSAONegotiator object at 0x28d4aee90>, context=None), 
'07Gr@1': NegotiatorInfo(negotiator=<negmas.sao.negotiators.controlled.ControlledSAONegotiator object at 0x28d4ad660>, context=None), 
'08SAA@1': NegotiatorInfo(negotiator=<negmas.sao.negotiators.controlled.ControlledSAONegotiator object at 0x28d4aec20>, context=None), 
'09SAA@1': NegotiatorInfo(negotiator=<negmas.sao.negotiators.controlled.ControlledSAONegotiator object at 0x28d4ac100>, context=None), 
'10Gr@1': NegotiatorInfo(negotiator=<negmas.sao.negotiators.controlled.ControlledSAONegotiator object at 0x28d4ac2b0>, context=None)}
supplier: [['SELLER'], ['00Ag@0', '01Gr@0', '02SAA@0', '03Gr@0', '04Gr@0'], ['05SAS@1', '06Gr@1', '07Gr@1', '08SAA@1', '09SAA@1', '10Gr@1']] 
consumer: [['00Ag@0', '01Gr@0', '02SAA@0', '03Gr@0', '04Gr@0'], ['05SAS@1', '06Gr@1', '07Gr@1', '08SAA@1', '09SAA@1', '10Gr@1'], ['BUYER']]
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
...
File ".../agent.py", line 30, in <dictcomp>
    partner: self.get_nmi(partner).issues[QUANTITY] for partner in partners
  File ".../scml/oneshot/agent.py", line 271, in get_nmi
    return self.negotiators[partner_id][0].nmi
KeyError: '06Gr@1'
```

According to the output, my agent is on the first layer, and `06Gr@1` is on the second layer. But `06Gr@1` is disappeared from `self.negotiators`, so I can't get the NMI of `06Gr@1`. I can't understand this behavior.

Is there any workaround?

**Comment by yasserfarouk** [comment]
Thank you for reporting this. We will investigate the issue and return to you as soon as possible. If you can **privately** share the complete agent code that generated the error, it will be faster. You can - if you are ok with it - send it to the organizer’s email.

As a first guess, it may be the case that some agents fail to enter the negotiation at all. In this case, you should not find their ifs in `self.active_negotiators`

**Comment by yasserfarouk** [comment]
Hi again @a4rcvv 

Could not reproduce this issue yet. I can see two possible explanations:
1. `06Gr@1` did not enter the negotiation (for example, it returned `None` from its `first_proposals` corresponding to your agent or ended the negotiation in the first round (by returning `END_NEGOTIATION` from its `counter` function.
2. `06Gr@1` was not invited to negotiate at all. This should only happen if it is bankrupt (has a negative balance) but that almost never happens in SCMLOneShot.

In both cases, you will not be able to access the NMI because it does not exist. In this case you can simply return `None` for this negotiation from your `first_proposals` as it will not be used.
It will be helpful to share the code of your agent and whatever Gr06 is using as well (yasserfarouk@gmail.com). 
Why do you need to access the NMI? If the goal is to get the issues, then you can use `current_input_issues` if you are a seller and `current_output_issues` if you are a buyer (in your example, you are a seller and should use `current_input_issues` ).

*Note that all negotiations have the same issues and ranges of values for each issue*

**Comment by a4rcvv** [comment]
Thank you for your research. 

> It will be helpful to share the code of your agent and whatever Gr06 is using as well

I have send my agent via email. `06Gr@1` is `scml.GreedyOneShotAgent`.

> Why do you need to access the NMI? If the goal is to get the issues, then you can use current_input_issues if you are a seller and current_output_issues if you are a buyer (in your example, you are a seller and should use current_input_issues ).
Note that all negotiations have the same issues and ranges of values for each issue

Exactly, I want to access the NMI to get the issues. So for now my agent is working well by using `self.awi.current_output_issues` and `self.awi.current_input_issues`. Thank you for telling this variable.

**Comment by a4rcvv** [comment]
> If the goal is to get the issues, then you can use `current_input_issues` if you are a seller and `current_output_issues` if you are a buyer 

By the way, I think we have to use `current_output_issues` (not `input`) if the agent is a seller, because it negotiates about its output products.

**Reply by yasserfarouk** [reply]
You are right. I always confuse these two :-)

will check your agent and report about the original issue asap

**Comment by yasserfarouk** [comment, marked as answer]
To recreate the environment we ran:

```sh
python3.10 -m venv .venv
source .venv/bin/activate
pip install scml scml-agents
```
Could not use your poetry.lock file because we have some issues with poetry.

I added the following assertions to test for the error you are experiencing in `first_proposals()`

```python
        received_partners = set(_ for _ in self.negotiators.keys())
        expected_partners = set(self.partners())
        assert (
            expected_partners == received_partners
        ), f"Expected partners: {expected_partners}\nFound partners {received_partners}"
```

The following test passes normally (it does take a very long time as we try the world 200 times for 100 simulation steps each):

```python
def test_tryrun():
    # SAA is single agreement agent
    buyers = ["SAS", "Greedy", "Greedy", "SAA", "SAA", "Greedy"]
    sellers = ["Neko23", "Greedy", "SAA", "Greedy", "Greedy"]
    agent_processes = [0] * len(sellers) + [1] * len(buyers)
    seller_types: list[type[OneShotAgent]] = get_agent_types(sellers)
    buyer_types: list[type[OneShotAgent]] = get_agent_types(buyers)
    for _ in range(200):
        # run the world
        world = SCML2023OneShotWorld(
            **SCML2023OneShotWorld.generate(
                agent_types=seller_types + buyer_types,
                agent_processes=agent_processes,
                n_processes=2,
                n_steps=100,
                random_agent_types=False,
            )
        )
        world.run()
```

This seems to reproduce your issue. We confirmed that the agents missing from the `self.negotiators` did indeed return `None` from their `proposed()` method which means they ended the negotiation before it even started. This may not be a rational behavior from these agents.

The most important guarantees are the following:
- The set of negotiators in `self.negotiators` is always a subset of `my_suppliers` or `my_consumers` appropriately in `first_proposals` and `counter_all`.
- In `counter_all` the set of negotiators in `states` matches `self.negotiators`.

Please let me know if that resolves your issue. We will keep investigating the cause of this immediate end of negotiation from some agents.

**Comment by yasserfarouk** [comment]
It seems that the problem stems from `GreedyOneShotAgent`. It estimates the demand in the market by dividing the minimum of exogenous contract quantities by the number of agents in its level. In your example, we have 6 buyers so if the minimum exogenous quantity is less than 6, it will conclude that it needs to buy nothing and will not enter any negotiations by returning `END_NEGOTIATION` from `respond()` and `None` from `propose()`.  That explains why you receive only a subset of negotiators in `first_proposal()` because the system will not bother asking you about the first proposal for an agent that already indicated it is leaving the negotiation. That is a feature not a bug.

**Reply by a4rcvv** [reply]
Thank you. So can we say the following?

- Keys of `self.negotiators` are the agent which are negotiating with me now.
- There is a case that other agents decide to end negotiations before `OneShotSyncAgent.first_proposal()` is called. So some agents (= agents which ended negotiations with me) may disappear from `self.negotiators`.

**Reply by yasserfarouk** [reply]
Exactly.

**Reply by a4rcvv** [reply]
I understood. Thank you very much.
